sql: error in cluster_logical_timestamp when not SERIALIZABLE

chrisseto opened the pull request:
Previously, running `cluster_logical_timestamp` within an isolation level that tolerated write skew would result in an unhandled panic.

This commit adds a gate at the SQL level that will instead return a FeatureNotSupported pgerror. This behavior may be permanent or may be removed once we determine how CommitTimestamp should function at these isolation levels.

Epic:  CRDB-26546
Fixes: #103245
Release note (sql change): cluster_logical_timestamp now returns an error when called at isolation levels lower than SERIALIZABLE.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107090)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
very nice! just some minor comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107090#-:-N_eYsLr11FHvx-0QnYT:bmijovs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto)

___
*[`pkg/sql/sem/eval/context.go` line 543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107090#-N_eXwne0ftwHSLhzHnI:-N_eXwne0ftwHSLhzHnJ:b-x81jqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/218d71e75400310a341bcb0b4da9034b7aee7842/pkg/sql/sem/eval/context.go#L543)):*
> ```Go
> 	// a pgerror until we decide how this will officially behave. See #103245.
> 	if ec.TxnIsoLevel.ToleratesWriteSkew() {
> 		return nil, pgerror.Newf(pgcode.FeatureNotSupported, "unsupported in %s isolation", ec.TxnIsoLevel.String())
> ```

super nit: could you use `kvTxnIsolationLevelToTree(ec.TxnIsoLevel).String()`. we have this bit of translation to keep the user-facing grammar separate from the internals of KV, and it has slightly more familiar string formatting.
___
*[`pkg/sql/logictest/testdata/logic_test/txn` line 1536 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107090#-N_eYnZyCPu5dRMOB5OQ:-N_eYnZyCPu5dRMOB5OR:b-1dkqx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/218d71e75400310a341bcb0b4da9034b7aee7842/pkg/sql/logictest/testdata/logic_test/txn#L1536)):*
> ```
> statement error pq: cluster_logical_timestamp\(\): unsupported in ReadCommitted isolation
> BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
> SELECT cluster_logical_timestamp();
> ```

nit: this should have a
```
statement ok
ROLLBACK
```

after this


<!-- Sent from Reviewable.io -->

chrisseto commented:
@rafiss could you point me in the right direction WRT the failing logic test? Seems like the test might be flake depending on which node you're executing against? If there's a way to read back the actual isolation level, we could skip the test if it isn't running at READ COMMITTED or we could gate the functionality on a specific version?

```
[20:06:40] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12659/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-mixed-22.2-23.1/local-mixed-22_2-23_1_test_/local-mixed-22_2-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/txn with config local-mixed-22.2-23.1: 371 tests, 0 failures
    logic.go:4068:
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12659/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-mixed-22.2-23.1/local-mixed-22_2-23_1_test_/local-mixed-22_2-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/txn:1534: error while processing
    logic.go:4068:
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12659/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-mixed-22.2-23.1/local-mixed-22_2-23_1_test_/local-mixed-22_2-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/txn:1534:
        expected "pq: cluster_logical_timestamp\\(\\): unsupported in READ COMMITTED isolation", but no error occurred
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c2c63d472fc75a33623ecf6091cebe9e/logTestLogic_txn879699003
```

rafiss commented:
Yes I can help for sure. The issue with this failure is:
```
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12659/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-mixed-22.2-23.1/local-mixed-22_2-23_1_test_/local-mixed-22_2-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/txn:1534: error while processing
    logic.go:4068:
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/12659/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local-mixed-22.2-23.1/local-mixed-22_2-23_1_test_/local-mixed-22_2-23_1_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/txn:1534:
        expected "pq: cluster_logical_timestamp\\(\\): unsupported in READ COMMITTED isolation", but no error occurred
```

To interpret that a bit, this is saying that the `txn` is failing under the `local-mixed-22_2-23_1` logictest configuration. That logictest configuration is defined here: https://github.com/cockroachdb/cockroach/blob/4618dbf07f1fe42e26e16c95b1b495119ad5d267/pkg/sql/logictest/logictestbase/logictestbase.go#L471-L478

The test configuration uses a "current" binary, but a "logical cluster version" from the previous release. This is so that it tests the mixed-version state of a CRDB cluster, which occurs during the process of an online upgrade. One of the things that we did for the READ COMMITTED project is we added a version gate so that READ COMMITTED can only be used if the cluster upgrade was finalized, and the cluster is no longer in a mixed version state. Otherwise, it falls back to the old behavior of automatically promoting READ COMMITTED to the SERIALIZABLE level. That gate is in this function: https://github.com/cockroachdb/cockroach/blob/4618dbf07f1fe42e26e16c95b1b495119ad5d267/pkg/sql/conn_executor.go#L3389

To put this all together, the issue you are seeing where no error occurs is actually expected under the `local-mixed-22_2-23_1` configuration. So the right thing to do in this case is to use the `skipif config local-mixed-22_2-23_1` directive. Check out the logictest docs for more information on `skipif`: https://github.com/cockroachdb/cockroach/blob/4618dbf07f1fe42e26e16c95b1b495119ad5d267/pkg/sql/logictest/logic.go#L392

chrisseto committed:
sql: error in cluster_logical_timestamp when not SERIALIZABLE

Previously, running `cluster_logical_timestamp` within an isolation
level that tolerated write skew would result in an unhandled panic.

This commit adds a gate at the SQL level that will instead return a
FeatureNotSupported pgerror. This behavior may be permanent or may be
removed once we determine how CommitTimestamp should function at these
isolation levels.

Epic:  CRDB-26546
Part of: #103245
Release note (sql change): cluster_logical_timestamp now returns an
error when called at isolation levels lower than SERIALIZABLE.

chrisseto commented:
Thanks! This round of CI should pass now 🤞

rafiss reviewed (approved):
nice!

chrisseto commented:
bors r+
